demos/nonlinear_QG_winddrivengyre/qg_winddrivengyre.py.rst

```diff
@@ -0,0 +1,279 @@
+
+The above problem is the weak form of the nonlinear Stommel problem.  The linear term arises from neglecting the nonlinear advection, and can easily be obtained by neglecting the first term on the left hand side.
+	  
+Defining the Problem
+====================
+
+Now that we know the weak form we are now ready to solve this using Firedrake!
+
+First, we import the Firedrake, PETSc, NumPy and UFL packages, ::
+
+  from firedrake import *
+  from firedrake.petsc import PETSc
+  import numpy as np
+  import ufl
+
+Next, we can define the geometry of our domain. In this example, we
+will be using a square of length one with 50 cells. ::
+  
+  n0 = 50           # Spatial resolution
+  Ly = 1.0          # Meridonal length
+  Lx = 1.0          # Zonal length
+  mesh = RectangleMesh(n0, n0, Lx, Ly, reorder = None)
+
+We can then define the Function Space within which the
+solution of the streamfunction will reside. ::
+
+  Vcg = FunctionSpace(mesh, 'CG', 3) # CG elements for Streamfunction
+
+We will also impose no-normal flow strongly to ensure that the
+boundary condition :math:`\psi = 0` will be met, ::
+  
+  bc = DirichletBC(Vcg, 0.0, 'on_boundary')
+
+Now we will define all the parameters we are using in this tutorial. ::
+
+  beta = Constant('1.0')      # Beta parameter
+  F = Constant('1.0')         # Burger number
+  r = Constant('0.2')         # Bottom drag
+  tau = Constant('0.001')     # Wind Forcing
+  x = SpatialCoordinate(mesh)
+  Qwinds = Function(Vcg).interpolate(-tau * cos(pi * (x[1]/Ly - 0.5)))
+
+We can now define the Test Function and the Trial Function of this problem, both must be in the same function space::
+
+  phi, psi = TestFunction(Vcg), TrialFunction(Vcg)
+
+We must define functions that will store our linear and nonlinear solutions.
+In order to solve the nonlinear problem, we use the linear
+solution as a guess for the nonlinear problem. ::
+
+  psi_lin = Function(Vcg, name='Linear Streamfunction')
+  psi_non = Function(Vcg, name='Nonlinear Streamfunction')
+
+We can finally write down the linear Stommel equation in its weak
+form. We will use the solution to this as the input for the nonlinear
+Stommel equation. ::
+
+  a = - r * inner(grad(psi), grad(phi)) * dx - F * psi * phi * dx + beta * psi.dx(0) * phi * dx
+  L = Qwinds * phi * dx
+
+We set-up an elliptic solver for this problem, and solve for the
+linear streamfunction, ::
+
+  linear_problem = LinearVariationalProblem(a, L, psi_lin, bcs=bc)
+  linear_solver = LinearVariationalSolver(linear_problem,
+                                          solver_parameters= {'ksp_type': 'preonly',
+                                                              'pc_type': 'lu'})
+  linear_solver.solve()
+
+We will redefine the nonlinear stream-function as its guess, the
+linear stream-function ::
```

Comment:
"We use the solution to the linear problem as an initial guess for the nonlinear one"?